judge/views/problem.py

```diff
@@ -201,6 +202,90 @@ def get_context_data(self, **kwargs):
+            except ObjectDoesNotExist:
+                context['vote'] = None
+        else:
+            context['vote'] = None
+
+        return context
+
+
+class ProblemVote(ProblemMixin, DetailView):
+    context_object_name = 'problem'
+    template_name = 'problem/vote-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        if self.object.get_vote_perm(self.request.user) < 2:
+            raise Http404()
+
+        context = super().get_context_data(**kwargs)
+
+        try:
+            context['vote'] = ProblemPointsVote.objects.get(voter=self.request.profile, problem=self.object)
+        except ObjectDoesNotExist:
+            context['vote'] = None
+
+        context['max_possible_vote'] = settings.DMOJ_PROBLEM_MAX_USER_POINTS_VOTE
+        context['min_possible_vote'] = max(settings.DMOJ_PROBLEM_MIN_USER_POINTS_VOTE,
+                                           settings.DMOJ_PROBLEM_MIN_PROBLEM_POINTS)
+        return context
+
+    def post(self, request, *args, **kwargs):
+        problem = self.get_object()
+        if problem.get_vote_perm(request.user) < 2:
+            return HttpResponseForbidden(_('Not allowed to vote on this problem.'), content_type='text/plain')
+
+        form = ProblemPointsVoteForm(request.POST)
+        if not form.is_valid():
+            return JsonResponse(form.errors, status=400)
+
+        with transaction.atomic():
+            # Delete any pre-existing votes.
+            ProblemPointsVote.objects.filter(voter=request.profile, problem=problem).delete()
+            vote = form.save(commit=False)
+            vote.voter = request.profile
+            vote.problem = problem
+            vote.save()
+        return JsonResponse({'points': vote.points})
+
+
+class DeleteProblemVote(ProblemMixin, SingleObjectMixin, View):
+    http_method_names = ['options', 'post']  # This disables GET requests, even though ProblemMixin.get exists.
+
+    def post(self, request, *args, **kwargs):
+        problem = self.get_object()
+        if problem.get_vote_perm(request.user) < 2:
+            return HttpResponseForbidden(_('Not allowed to delete votes on this problem.'), content_type='text/plain')
+
+        ProblemPointsVote.objects.filter(voter=request.profile, problem=problem).delete()
+        return HttpResponse('success', content_type='text/plain')
+
+
+class ProblemVoteStats(ProblemMixin, DetailView):
+    context_object_name = 'problem'
+    template_name = 'problem/vote-stats-ajax.html'
+
+    def get_context_data(self, **kwargs):
+        if self.object.get_vote_perm(self.request.user) < 1:
+            raise Http404()
+
+        context = super().get_context_data(**kwargs)
+
+        votes = list(self.object.problem_points_votes.order_by('points').values_list('points', flat=True))
+        context['votes'] = votes
+
+        if votes:
+            context['mean'] = sum(votes) / len(votes)
+            context['median'] = median(votes)
```

Comment:
Nit: import `statistics` and use both `mean` and `median`. This makes it more clear that we're using a module, and there's no reason to split stdlib and hand-rolled.

Reply:
done

Reply:
I meant:

```py
import statistics

context['mean'] = statistics.mean(votes)
context['median'] = statistics.median(votes)
```

This makes it clear where the functions are coming from.

Reply:
Leaving for a future reviewer.

Reply:
for context, this change wouldn't help me

![Capture](https://user-images.githubusercontent.com/14223529/189497563-53db90ff-a379-4e5a-aeac-56574a080816.PNG)

Reply:
Since the function is well known and only really has one interpretation, it doesn't matter too much where it comes from. I think we can just keep it as-is.